src/Workspaces/Core/Portable/Utilities/OwnedDisposable`1.cs

```diff
@@ -0,0 +1,101 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+
+namespace Roslyn.Utilities
+{
+    internal struct OwnedDisposable<T> : IDisposable, IEquatable<OwnedDisposable<T>>
+        where T : class, IDisposable
+    {
+        private T _resource;
+
+        public OwnedDisposable(T resource)
+        {
+            _resource = resource;
+        }
+
+        public OwnedDisposable(Func<T> resource)
+        {
```

Comment:
I am not sure when this ctor should be used compared to the one above.
> The factory function avoids the need to transfer an object into this constructor
is there reason you want to avoid transfer ownership? and it is using ref, so does it require to move ownership? a bit hard to know when I should and shouldn't move ownership.

Reply:
> I am not sure when this ctor should be used compared to the one above.

The previous constructor avoids an allocation, but I haven't found a way to use it without triggering one of the `IDisposable` analyzer warnings. A key design goal of this type and usage is addressing https://github.com/dotnet/roslyn/issues/25880#issuecomment-377967573 in a world where the following hold:

* The `IDisposable` analyzers are enabled with severity Error
* Warnings for the analyzers must not be suppressed

> is there reason you want to avoid transfer ownership? and it is using ref, so does it require to move ownership? a bit hard to know when I should and shouldn't move ownership. it would be nicer if one can more easily know when to call Claim or Move or use ref.

These rules are not fully created yet. The answers will depend on how the previous item drives the design.